inc/saipoe.h

```diff
@@ -0,0 +1,522 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saipoe.h
+ *
+ * @brief   This module defines SAI POE interface
+ */
+
+#if !defined (__SAIPOE_H_)
+#define __SAIPOE_H_
+
+#include <saitypes.h>
+
+/**
+ * @brief POE device power limit mode
+ *
+ * port limit - max power limit per port is configured by the user
+ * class limit - max power is set automatically according the class of the connected device
+ */
+typedef enum _sai_poe_device_limit_mode_t
+{
+    SAI_POE_DEVICE_LIMIT_MODE_PORT  = 0,
+    SAI_POE_DEVICE_LIMIT_MODE_CLASS  = 1
+} sai_poe_device_limit_mode_t;
+
+/**
+ * @brief POE device PSE (power sourcing equipment) status
+ */
+typedef enum _sai_poe_pse_status_t
+{
+    SAI_POE_PSE_STATUS_TYPE_ACTIVE,
+    SAI_POE_PSE_STATUS_TYPE_FAIL,
+    SAI_POE_PSE_STATUS_TYPE_NOT_PRESENT,
+} sai_poe_pse_status_t;
+
+/**
+ * @brief POE port (IEEE) standard
+ */
+typedef enum _sai_poe_port_standard_t
+{
+    SAI_POE_PORT_STANDARD_TYPE_AF,
+    SAI_POE_PORT_STANDARD_TYPE_AT,
+    SAI_POE_PORT_STANDARD_TYPE_60W,
+    SAI_POE_PORT_STANDARD_TYPE_BT_TYPE3,
+    SAI_POE_PORT_STANDARD_TYPE_BT_TYPE4,
+} sai_poe_port_standard_t;
+
+/**
+ * @brief POE port power priority
+ */
+typedef enum _sai_poe_port_power_priority_t
+{
+    SAI_POE_PORT_POWER_PRIORITY_TYPE_LOW,
+    SAI_POE_PORT_POWER_PRIORITY_TYPE_HIGH,
+    SAI_POE_PORT_POWER_PRIORITY_TYPE_CRITICAL,
+} sai_poe_port_power_priority_t;
+
+/**
+ * @brief POE port status
+ */
+typedef enum _sai_poe_port_status_t
+{
+    SAI_POE_PORT_STATUS_TYPE_OFF,
+    SAI_POE_PORT_STATUS_TYPE_SEARCHING,
+    SAI_POE_PORT_STATUS_TYPE_DELIVERING_POWER,
+    SAI_POE_PORT_STATUS_TYPE_FAULT,
+} sai_poe_port_status_t;
+
+/**
+ * @brief POE device attributes
+ */
+typedef enum _sai_poe_device_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_POE_DEVICE_ATTR_START,
+
+    /**
+     * @brief Device Information for POE initialization.
+     *
+     * Hardware information format is based on POE SAI implementations by vendors.
+     * Format is vendor specific.
+     * Example: Like PCI location, I2C address, UART etc.
+     *
+     * @type char
```

Comment:
this will point to sai_attribute_value_t.chardata[32] type, is that what you want ?

Reply:
Yes, a string allows each vendor the flexibility to pass whatever information they need on how to communicate with the PoE hardware.
For example, if the board has a MCU you can set the hw_info to "integrated_mcu". If the board uses direct i2c access then you can set the hw_info to "i2c" or an address. You can also send an empty string if this is not relevant for the system.

Reply:
A simple string will not be sufficient. Better define a model for possible interfaces.
1) UART :
    Dev name/node, baud rate, parity settings etc
2) I2C:
    Bus number, i2C address and the mode of I2C/SMBUS transaction etc
3) SPI:
    SPI address, CS, Cpol, Cphase etc
4) PCI or shared memory:
    Memory start, Memory size, R/W attributes, etc
and so on.